### PR TITLE
Allow arbitrary offset for GEP

### PR DESCRIPTION
**Pull request by lialan**

### CODE REVIEW

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Target/SyncVM/SyncVMCodegenPrepare.cpp [hunk 1]
```diff
@@ -148,8 +148,8 @@ bool SyncVMCodegenPrepare::convertPointerArithmetics(Function &F) {
     Value *TmpOffset = Builder.getIntN(256, 0);
     APInt Offset(256, 0);
     if (GEP->accumulateConstantOffset(*DL, Offset)) {
-      // the 32 here is the word length for zkEVM
-      TmpOffset = Builder.getIntN(256, Offset.getZExtValue());
+      // We allow arbitrary GEP offset (256-bit)
+      TmpOffset = Builder.getInt(Offset);
```
Comment (akiramenai) [inline llvm/lib/Target/SyncVM/SyncVMCodegenPrepare.cpp]: My personal solution was
```
TmpOffset = Builder.getInt(Offset.sext(256));
```
Please add a test which trigger offset addition with different bitness of APInts.

**Review by akiramenai — CHANGES_REQUESTED** [review]
Tests are missing

File: llvm/lib/Target/SyncVM/SyncVMCodegenPrepare.cpp (hunk 1 above)
Comment (lialan) [reply llvm/lib/Target/SyncVM/SyncVMCodegenPrepare.cpp]: Updated

**Comment by lialan** [comment]
Added the test @akiramenai mentioned and another 128bit test.

**Review by akiramenai — APPROVED** [review]

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

**Review by hedgar2017 — APPROVED** [review]

File: llvm/lib/Target/SyncVM/SyncVMCodegenPrepare.cpp [hunk 2]
```diff
@@ -148,8 +148,8 @@ bool SyncVMCodegenPrepare::convertPointerArithmetics(Function &F) {
     Value *TmpOffset = Builder.getIntN(256, 0);
     APInt Offset(256, 0);
     if (GEP->accumulateConstantOffset(*DL, Offset)) {
-      // the 32 here is the word length for zkEVM
-      TmpOffset = Builder.getIntN(256, Offset.getZExtValue());
+      // We allow arbitrary GEP offset (256-bit)
+      TmpOffset = Builder.getInt(Offset.sext(256));
```
Comment (hedgar2017) [inline llvm/lib/Target/SyncVM/SyncVMCodegenPrepare.cpp:152]: sext?

**Review by hedgar2017 — CHANGES_REQUESTED** [review]

**Review by hedgar2017 — APPROVED** [review]

**Pull request merged**